lxd/storage/drivers/driver_zfs_volumes.go

```diff
@@ -2609,38 +2609,33 @@ func (d *zfs) MountVolumeSnapshot(snapVol Volume, op *operations.Operation) erro
 	unlock := snapVol.MountLock()
 	defer unlock()
 
-	var err error
 	mountPath := snapVol.MountPath()
 	snapshotDataset := d.dataset(snapVol, false)
 
 	revert := revert.New()
 	defer revert.Fail()
 
 	// Check if filesystem volume already mounted.
-	if snapVol.contentType == ContentTypeFS && !filesystem.IsMountPoint(mountPath) {
-		err := snapVol.EnsureMountPath()
-		if err != nil {
-			return err
-		}
+	if snapVol.contentType == ContentTypeFS && !d.isBlockBacking(snapVol) {
+		if !filesystem.IsMountPoint(mountPath) {
+			err := snapVol.EnsureMountPath()
+			if err != nil {
+				return err
+			}
 
-		// Mount the snapshot directly (not possible through tools).
-		err = TryMount(snapshotDataset, mountPath, "zfs", 0, "")
-		if err != nil {
-			return err
-		}
+			// Mount the snapshot directly (not possible through tools).
+			err = TryMount(snapshotDataset, mountPath, "zfs", 0, "")
+			if err != nil {
+				return err
+			}
 
-		d.logger.Debug("Mounted ZFS snapshot dataset", logger.Ctx{"dev": snapshotDataset, "path": mountPath})
-	} else if snapVol.contentType == ContentTypeBlock {
+			d.logger.Debug("Mounted ZFS snapshot dataset", logger.Ctx{"dev": snapshotDataset, "path": mountPath})
+		}
+	} else {
 		// For block devices, we make them appear by enabling volmode=dev and snapdev=visible on the parent volume.
 		// Ensure snap volume parent is activated to avoid issues activating the snapshot volume device.
 		parent, _, _ := api.GetParentAndSnapshotName(snapVol.Name())
 		parentVol := NewVolume(d, d.Name(), snapVol.volType, snapVol.contentType, parent, snapVol.config, snapVol.poolConfig)
-		err = d.MountVolume(parentVol, op)
```

Comment:
Why did this bit need to go?
Also there is no equivalent defer that is unsetting volmode on the parent. The use of `Unmount` would have been handling that safely.

Reply:
XFS would complain if trying to mount the snapshot while its parent was mounted. The error message was `invalid argument` (I think). Not very helpful.

Reply:
Ah yes, well what you need is this, as the parent may be already mounted by another user/operation and you cannot unmount, so you need to do this instead:

https://github.com/lxc/lxd/blob/master/lxd/storage/drivers/driver_lvm_volumes.go#L936-L967